src/backgrounddownloader.cpp

```diff
@@ -0,0 +1,152 @@
+#include "backgrounddownloader.h"
+
+#include <iostream>
+#include <QString>
+#include <QTimer>
+#include <QDebug>
+
+BackgroundDownloader::BackgroundDownloader(kiwix::Downloader* downloader)
+    :  mp_downloader(downloader)
+{
+    QThread* newThread = new QThread(); // no parent
+    m_thread.reset(newThread);
+    moveToThread(m_thread.get());
+
+    // the timer causes updateStatus() to be called once per second on this thread's event loop
+    mp_timer = new QTimer(newThread);
+    connect(mp_timer, SIGNAL(timeout()), this, SLOT(updateStatus()));
+    mp_timer->start(1000);
+
+    m_thread->start();
+}
+
+BackgroundDownloader::~BackgroundDownloader() {
+    QMetaObject::invokeMethod(this, "cleanup");
+    m_thread->wait();
+}
+
+void BackgroundDownloader::cleanup() {
+    // TODO: delete resources
+
+    m_thread->quit();
+}
+
+// getDownloadStatus returns the status map for this downloadId. May be called from a separate thread
+QMap<std::string, std::string> BackgroundDownloader::getDownloadStatus(const std::string did)
```

Comment:
`getDownloadStatus()` is called by `ContentManager:: updateDownloadInfos()`, which is running on the UI thread. `m_rwlock` protects the map read.